File: 2022/day21/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Solution {

    private Map<String, Object> monkeys = new HashMap<>();
    private Map<String, Long> evalMemo = new HashMap<>();
    private Map<String, Boolean> humnMemo = new HashMap<>();

    public static void main(String[] args) throws IOException {
        Solution solution = new Solution();
        String inputPath = Paths.get(System.getProperty("user.dir"), "..", "input.txt")
                                .normalize().toString();
        String text = Files.readString(Path.of(inputPath));

        System.out.println("Part 1: " + solution.part1(text));
        System.out.println("Part 2: " + solution.part2(text));
    }

    private void parseInput(String text) {
        monkeys.clear();
        evalMemo.clear();
        humnMemo.clear();

        for (String line : text.trim().split("\n")) {
            String[] parts = line.split(": ");
            String name = parts[0];
            String job = parts[1];
            String[] jobParts = job.split(" ");

            if (jobParts.length == 1) {
                monkeys.put(name, Long.parseLong(jobParts[0]));
            } else {
                monkeys.put(name, new String[]{jobParts[0], jobParts[1], jobParts[2]});
            }
        }
    }

    private long evaluate(String name) {
        if (evalMemo.containsKey(name)) {
            return evalMemo.get(name);
        }

        Object job = monkeys.get(name);
        if (job instanceof Long) {
            return (Long) job;
        }

        String[] parts = (String[]) job;
        String left = parts[0];
        String op = parts[1];
        String right = parts[2];

        long leftVal = evaluate(left);
        long rightVal = evaluate(right);

        long result;
        switch (op) {
            case "+": result = leftVal + rightVal; break;
            case "-": result = leftVal - rightVal; break;
            case "*": result = leftVal * rightVal; break;
            case "/": result = leftVal / rightVal; break;
            default: throw new IllegalArgumentException("Unknown operator: " + op);
        }

        evalMemo.put(name, result);
        return result;
    }

    private boolean containsHumn(String name) {
        if (humnMemo.containsKey(name)) {
            return humnMemo.get(name);
        }

        if (name.equals("humn")) {
            return true;
        }

        Object job = monkeys.get(name);
        if (job instanceof Long) {
            humnMemo.put(name, false);
            return false;
        }

        String[] parts = (String[]) job;
        boolean result = containsHumn(parts[0]) || containsHumn(parts[2]);
        humnMemo.put(name, result);
        return result;
    }

    private long solveForHumn(String name, long target) {
        if (name.equals("humn")) {
            return target;
        }

        Object job = monkeys.get(name);
        if (job instanceof Long) {
            throw new IllegalStateException("Cannot solve for humn through a literal number");
        }

        String[] parts = (String[]) job;
        String left = parts[0];
        String op = parts[1];
        String right = parts[2];

        boolean leftHasHumn = containsHumn(left);

        if (leftHasHumn) {
            long rightVal = evaluate(right);
            long newTarget;
            switch (op) {
                case "+": newTarget = target - rightVal; break;
                case "-": newTarget = target + rightVal; break;
                case "*": newTarget = target / rightVal; break;
                case "/": newTarget = target * rightVal; break;
                default: throw new IllegalArgumentException("Unknown operator: " + op);
            }
            return solveForHumn(left, newTarget);
        } else {
            long leftVal = evaluate(left);
            long newTarget;
            switch (op) {
                case "+": newTarget = target - leftVal; break;
                case "-": newTarget = leftVal - target; break;
                case "*": newTarget = target / leftVal; break;
                case "/": newTarget = leftVal / target; break;
                default: throw new IllegalArgumentException("Unknown operator: " + op);
            }
            return solveForHumn(right, newTarget);
        }
    }

    public long part1(String text) {
        parseInput(text);
        return evaluate("root");
    }

    public long part2(String text) {
        parseInput(text);

        String[] rootJob = (String[]) monkeys.get("root");
        String left = rootJob[0];
        String right = rootJob[2];

        boolean leftHasHumn = containsHumn(left);

        if (leftHasHumn) {
            long target = evaluate(right);
            return solveForHumn(left, target);
        } else {
            long target = evaluate(left);
            return solveForHumn(right, target);
        }
    }
}

File: 2022/day21/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

// Monkey represents either a number or an operation
type Monkey struct {
	isNumber bool
	value    int64
	left     string
	op       string
	right    string
}

func parseInput(text string) map[string]Monkey {
	monkeys := make(map[string]Monkey)
	scanner := bufio.NewScanner(strings.NewReader(text))

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		parts := strings.SplitN(line, ": ", 2)
		name := parts[0]
		job := parts[1]

		jobParts := strings.Fields(job)
		if len(jobParts) == 1 {
			val, _ := strconv.ParseInt(jobParts[0], 10, 64)
			monkeys[name] = Monkey{isNumber: true, value: val}
		} else {
			monkeys[name] = Monkey{
				isNumber: false,
				left:     jobParts[0],
				op:       jobParts[1],
				right:    jobParts[2],
			}
		}
	}

	return monkeys
}

func evaluate(monkeys map[string]Monkey, name string, memo map[string]int64) int64 {
	if val, ok := memo[name]; ok {
		return val
	}

	monkey := monkeys[name]
	if monkey.isNumber {
		return monkey.value
	}

	leftVal := evaluate(monkeys, monkey.left, memo)
	rightVal := evaluate(monkeys, monkey.right, memo)

	var result int64
	switch monkey.op {
	case "+":
		result = leftVal + rightVal
	case "-":
		result = leftVal - rightVal
	case "*":
		result = leftVal * rightVal
	case "/":
		result = leftVal / rightVal
	}

	memo[name] = result
	return result
}

func containsHumn(monkeys map[string]Monkey, name string, memo map[string]bool) bool {
	if val, ok := memo[name]; ok {
		return val
	}

	if name == "humn" {
		return true
	}

	monkey := monkeys[name]
	if monkey.isNumber {
		memo[name] = false
		return false
	}

	result := containsHumn(monkeys, monkey.left, memo) || containsHumn(monkeys, monkey.right, memo)
	memo[name] = result
	return result
}

func solveForHumn(monkeys map[string]Monkey, name string, target int64, humnMemo map[string]bool, evalMemo map[string]int64) int64 {
	if name == "humn" {
		return target
	}

	monkey := monkeys[name]
	if monkey.isNumber {
		// Can't solve if it's just a number
		return 0
	}

	leftHasHumn := containsHumn(monkeys, monkey.left, humnMemo)

	if leftHasHumn {
		// Evaluate right side to get its value
		rightVal := evaluate(monkeys, monkey.right, evalMemo)
		var newTarget int64

		switch monkey.op {
		case "+":
			// left + right = target => left = target - right
			newTarget = target - rightVal
		case "-":
			// left - right = target => left = target + right
			newTarget = target + rightVal
		case "*":
			// left * right = target => left = target / right
			newTarget = target / rightVal
		case "/":
			// left / right = target => left = target * right
			newTarget = target * rightVal
		}
		return solveForHumn(monkeys, monkey.left, newTarget, humnMemo, evalMemo)
	} else {
		// Evaluate left side to get its value
		leftVal := evaluate(monkeys, monkey.left, evalMemo)
		var newTarget int64

		switch monkey.op {
		case "+":
			// left + right = target => right = target - left
			newTarget = target - leftVal
		case "-":
			// left - right = target => right = left - target
			newTarget = leftVal - target
		case "*":
			// left * right = target => right = target / left
			newTarget = target / leftVal
		case "/":
			// left / right = target => right = left / target
			newTarget = leftVal / target
		}
		return solveForHumn(monkeys, monkey.right, newTarget, humnMemo, evalMemo)
	}
}

func part1(text string) int64 {
	monkeys := parseInput(text)
	memo := make(map[string]int64)
	return evaluate(monkeys, "root", memo)
}

func part2(text string) int64 {
	monkeys := parseInput(text)

	// root checks equality between its two children
	root := monkeys["root"]
	left := root.left
	right := root.right

	humnMemo := make(map[string]bool)
	evalMemo := make(map[string]int64)

	// Find which side contains humn
	leftHasHumn := containsHumn(monkeys, left, humnMemo)

	if leftHasHumn {
		// Right side gives us the target value
		target := evaluate(monkeys, right, evalMemo)
		return solveForHumn(monkeys, left, target, humnMemo, evalMemo)
	} else {
		// Left side gives us the target value
		target := evaluate(monkeys, left, evalMemo)
		return solveForHumn(monkeys, right, target, humnMemo, evalMemo)
	}
}

func main() {
	_, filename, _, _ := runtime.Caller(0)
	inputPath := filepath.Join(filepath.Dir(filename), "..", "input.txt")

	content, err := os.ReadFile(inputPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	text := string(content)

	fmt.Println("Part 1:", part1(text))
	fmt.Println("Part 2:", part2(text))
}
